server/bucket.cc

```diff
@@ -0,0 +1,262 @@
+/*****************************************************************************/
+
+void
+WritableBucketList::
+init(size_t numElements, uint32_t numBuckets)
+{
+    this->numBuckets = numBuckets;
+    entryBits = ML::highest_bit(numBuckets) + 1;
+
+    // Take a number of bits per entry that evenly divides into
+    // 64 bits.
+    if (entryBits == 0) ;
+    else if (entryBits == 1) ;
+    else if (entryBits == 2) ;
+    else if (entryBits <= 4)
+        entryBits = 4;
+    else if (entryBits <= 8)
+        entryBits = 8;
+    else if (entryBits <= 16)
+        entryBits = 16;
+    else entryBits = 32;
+
+    //cerr << "using " << entryBits << " bits for " << numBuckets
+    //     << " buckets" << endl;
+
+    size_t numWords = (entryBits * numElements + 63) / 64;
+    auto writableStorage = makeSharedArray<uint64_t>(numWords);
+    this->current = writableStorage.get();
+    this->storage = writableStorage;
+    this->bitsWritten = 0;
+    this->numEntries = numElements;
+    this->numWritten = 0;
+}
+
+
+/*****************************************************************************/
+/* BUCKET DESCRIPTIONS                                                       */
+/*****************************************************************************/
+
+uint32_t
+BucketDescriptions::
+getBucket(const CellValue & val) const
+{
+    switch (val.cellType()) {
+    case CellValue::EMPTY:
+        if (!hasNulls)
+            throw HttpReturnException(500, "Unknown type for bucket");
+        return 0;
+
+    case CellValue::INTEGER:
+    case CellValue::FLOAT:
+        return numeric.getBucket(val.toDouble());
+    case CellValue::ASCII_STRING:
+    case CellValue::UTF8_STRING:
+        return strings.getBucket(val);
+    case CellValue::BLOB:
+        return blobs.getBucket(val);
+    case CellValue::TIMESTAMP:
+        return timestamps.getBucket(val);
+    case CellValue::TIMEINTERVAL:
+        return intervals.getBucket(val);
+
+    case CellValue::NUM_CELL_TYPES:
+        break;
+    }
+
+    throw HttpReturnException(500, "Unknown CellValue type for getBucket()");
+}
+
+uint32_t
+NumericValues::
+getBucket(double val) const
+{
+    if (!active)
+        throw HttpReturnException(500, "Attempt to get bucket from non-numeric value");
+    return std::lower_bound(splits.begin(), splits.end(), val)
+        - splits.begin() + offset;
+}
+
+uint32_t
+OrdinalValues::
+getBucket(const CellValue & val) const
+{
+    if (!active)
+        throw HttpReturnException(500, "Attempt to get bucket from non-ordinal value");
+    return std::lower_bound(splits.begin(), splits.end(), val)
+        - splits.begin() + offset;
+}
+
+uint32_t
+CategoricalValues::
+getBucket(const CellValue & val) const
+{
+    auto it = std::find(buckets.begin(), buckets.end(), val);
+    if (it != buckets.end())
+        return it - buckets.begin() + offset;
+    throw HttpReturnException(500, "categorical value " + val.toString() + " not found in col");
```

Comment:
The code often use single quotation when returning a value so that if the value is empty, the message is clearer.